File: java/examples/src/main/java/examples/prediction/PolymarketEndToEnd.java
package examples.prediction;

import io.github.ccxt.exchanges.prediction.Polymarket;
import io.github.ccxt.types.PredictionEvent;
import io.github.ccxt.types.PredictionMarket;
import io.github.ccxt.types.PredictionOrder;
import io.github.ccxt.types.PredictionOrderBook;
import io.github.ccxt.types.PredictionOutcome;
import io.github.ccxt.types.PredictionTicker;
import io.github.ccxt.types.PredictionTrade;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Polymarket end-to-end example (read market data + place/fetch/cancel one order).
 *
 * Flow:
 *   1. pick a high-volume event, a market inside it, and an outcome with a live two-sided book
 *   2. fetch the order book, ticker and recent trades for that outcome
 *   3. place a resting limit BUY well below the book, fetch it back, then cancel it
 *
 * Usage:
 *   POLYMARKET_PRIVATEKEY=... POLYMARKET_WALLETADDRESS=0x... \
 *   ./gradlew :examples:run -PmainClass=examples.prediction.PolymarketEndToEnd
 *
 * walletAddress is the polymarket account wallet (the proxy / deposit wallet shown in
 * your polymarket profile), privateKey is the key of the EOA that owns it.
 */
public class PolymarketEndToEnd {

    static final double MAX_NOTIONAL_USD = 25;  // hard cap per trade
    static final double ORDER_SIZE_SHARES = 5;  // polymarket minimum order size

    public static void main(String[] args) {
        String privateKey = System.getenv("POLYMARKET_PRIVATEKEY");
        String walletAddress = System.getenv("POLYMARKET_WALLETADDRESS");
        if (privateKey == null || walletAddress == null) {
            System.out.println("Set POLYMARKET_PRIVATEKEY and POLYMARKET_WALLETADDRESS env vars first.");
            return;
        }
        Map<String, Object> config = new HashMap<>();
        config.put("privateKey", privateKey);
        config.put("walletAddress", walletAddress);
        Polymarket exchange = new Polymarket(config);

        // 1) pick a high-volume event and an outcome with a live two-sided book ------------
        Map<String, Object> eventsParams = new HashMap<>();
        eventsParams.put("sort", "volume");
        eventsParams.put("limit", 15);
        List<PredictionEvent> events = exchange.fetchEvents(eventsParams);
        String symbol = null;
        PredictionOrderBook book = null;
        double tick = 0.01;
        int probes = 0;
        outer:
        for (PredictionEvent event : events) {
            if (event.markets == null) continue;
            for (PredictionMarket market : event.markets) {
                if (market.outcomes == null) continue;
                for (PredictionOutcome outcome : market.outcomes) {
                    if (probes >= 20 || outcome.outcome == null) {
                        break outer;
                    }
                    probes += 1;
                    PredictionOrderBook candidate = exchange.fetchOrderBook(outcome.outcome);
                    if (candidate.bids != null && !candidate.bids.isEmpty() && candidate.asks != null && !candidate.asks.isEmpty()) {
                        symbol = outcome.outcome;
                        book = candidate;
                        if (outcome.precision != null && outcome.precision.price != null) {
                            tick = outcome.precision.price;
                        }
                        System.out.println("event:    " + event.title);
                        System.out.println("outcome:  " + symbol + " (" + outcome.label + ")");
                        break outer;
                    }
                }
            }
        }
        if (symbol == null) {
            System.out.println("Could not find an outcome with a live two-sided order book right now.");
            return;
        }

        // 2) market data for the chosen outcome --------------------------------------------
        List<Double> bestBid = book.bids.get(0);
        List<Double> bestAsk = book.asks.get(0);
        System.out.println("\n--- market data ---");
        System.out.println("orderbook bid/ask: " + bestBid + " / " + bestAsk);
        try {
            PredictionTicker ticker = exchange.fetchTicker(symbol);
            System.out.println("ticker bid/ask/last: " + ticker.bid + " / " + ticker.ask + " / " + ticker.last);
        } catch (Exception e) {
            System.out.println("ticker:        n/a (" + e.getClass().getSimpleName() + ")");
        }
        try {
            List<PredictionTrade> trades = exchange.fetchTrades(symbol, null, 3L);
            System.out.println("recent trades: " + trades.size() + (trades.isEmpty() ? "" : (" last @ " + trades.get(0).price)));
        } catch (Exception e) {
            System.out.println("trades:        n/a (" + e.getClass().getSimpleName() + ")");
        }

        // 3) place a resting limit BUY well below the book, fetch it, then cancel -----------
        double bidPrice = bestBid.get(0);
        // half the best bid, floored to the tick — far below the ask, so it cannot fill
        double price = Math.floor((bidPrice * 0.5) / tick) * tick;
        price = Math.max(tick, Math.round(price * 10000.0) / 10000.0);
        double notional = ORDER_SIZE_SHARES * price;
        System.out.println("\n--- order ---");
        System.out.println("placing limit BUY " + ORDER_SIZE_SHARES + " shares @ " + price + " (notional " + (Math.round(notional * 100.0) / 100.0) + " USD)");
        if (notional >= MAX_NOTIONAL_USD) {
            System.out.println("ABORT: notional >= " + MAX_NOTIONAL_USD + " USD safety cap.");
            return;
        }

        PredictionOrder order = null;
        int exitCode = 0;
        try {
            order = exchange.createOrder(symbol, "limit", "buy", ORDER_SIZE_SHARES, price);
            System.out.println("placed:  id " + order.id + " | status " + order.status);
            PredictionOrder fetched = exchange.fetchOrder(order.id, symbol);
            System.out.println("fetched: id " + fetched.id + " | status " + fetched.status + " | remaining " + fetched.remaining);
        } catch (Exception e) {
            System.out.println("order flow failed: " + e);
            exitCode = 1;
        } finally {
            if (order != null && order.id != null) {
                PredictionOrder canceled = exchange.cancelOrder(order.id, symbol);
                System.out.println("canceled: id " + canceled.id + " | status " + canceled.status);
            }
        }
        System.exit(exitCode); // the exchange keeps a background thread pool alive
    }
}

File: examples/cs/examples/prediction/PolymarketEndToEnd.cs
using ccxt;

namespace examples;

// Polymarket end-to-end example (read market data + place/fetch/cancel one order).
//
// Flow:
//   1. pick a high-volume event, a market inside it, and an outcome with a live two-sided book
//   2. fetch the order book, ticker and recent trades for that outcome
//   3. place a resting limit BUY well below the book, fetch it back, then cancel it
//
// Usage:
//   POLYMARKET_PRIVATEKEY=... POLYMARKET_WALLETADDRESS=0x... dotnet run
//
// walletAddress is the polymarket account wallet (the proxy / deposit wallet shown in
// your polymarket profile), privateKey is the key of the EOA that owns it.

partial class Examples
{
    const double MaxNotionalUsd = 25;  // hard cap per trade
    const double OrderSizeShares = 5;  // polymarket minimum order size

    public static async Task PolymarketEndToEnd()
    {
        var privateKey = Environment.GetEnvironmentVariable("POLYMARKET_PRIVATEKEY");
        var walletAddress = Environment.GetEnvironmentVariable("POLYMARKET_WALLETADDRESS");
        if (string.IsNullOrEmpty(privateKey) || string.IsNullOrEmpty(walletAddress))
        {
            Console.WriteLine("Set POLYMARKET_PRIVATEKEY and POLYMARKET_WALLETADDRESS env vars first.");
            return;
        }
        var exchange = new ccxt.prediction.polymarket(new Dictionary<string, object>() {
            { "privateKey", privateKey },
            { "walletAddress", walletAddress },
        });

        // 1) pick a high-volume event and an outcome with a live two-sided book ------------
        var events = await exchange.FetchEvents(new Dictionary<string, object>() { { "sort", "volume" }, { "limit", 15 } });
        string symbol = null;
        PredictionOrderBook? chosenBook = null;
        double tick = 0.01;
        var probes = 0;
        foreach (var ev in events)
        {
            foreach (var market in ev.markets ?? new List<PredictionMarket>())
            {
                foreach (var outcome in market.outcomes ?? new List<PredictionOutcome>())
                {
                    if (probes >= 20 || outcome.outcome == null)
                    {
                        break;
                    }
                    probes += 1;
                    var book = await exchange.FetchOrderBook(outcome.outcome);
                    if ((book.bids?.Count ?? 0) > 0 && (book.asks?.Count ?? 0) > 0)
                    {
                        symbol = outcome.outcome;
                        chosenBook = book;
                        tick = outcome.precision?.price ?? 0.01;
                        Console.WriteLine("event:    " + ev.title);
                        Console.WriteLine("outcome:  " + symbol + " (" + outcome.label + ")");
                        break;
                    }
                }
                if (symbol != null) break;
            }
            if (symbol != null) break;
        }
        if (symbol == null)
        {
            Console.WriteLine("Could not find an outcome with a live two-sided order book right now.");
            return;
        }

        // 2) market data for the chosen outcome --------------------------------------------
        var bestBid = chosenBook.Value.bids[0];
        var bestAsk = chosenBook.Value.asks[0];
        Console.WriteLine("\n--- market data ---");
        Console.WriteLine("orderbook bid/ask: [" + bestBid[0] + ", " + bestBid[1] + "] / [" + bestAsk[0] + ", " + bestAsk[1] + "]");
        try
        {
            var ticker = await exchange.FetchTicker(symbol);
            Console.WriteLine("ticker bid/ask/last: " + ticker.bid + " / " + ticker.ask + " / " + ticker.last);
        }
        catch (Exception e) { Console.WriteLine("ticker:        n/a (" + e.GetType().Name + ")"); }
        try
        {
            var trades = await exchange.FetchTrades(symbol, null, 3);
            Console.WriteLine("recent trades: " + trades.Count + (trades.Count > 0 ? (" last @ " + trades[0].price) : ""));
        }
        catch (Exception e) { Console.WriteLine("trades:        n/a (" + e.GetType().Name + ")"); }

        // 3) place a resting limit BUY well below the book, fetch it, then cancel -----------
        var bidPrice = bestBid[0];
        // half the best bid, floored to the tick — far below the ask, so it cannot fill
        var price = Math.Max(tick, Math.Round(Math.Floor((bidPrice * 0.5) / tick) * tick, 4));
        var notional = OrderSizeShares * price;
        Console.WriteLine("\n--- order ---");
        Console.WriteLine("placing limit BUY " + OrderSizeShares + " shares @ " + price + " (notional " + Math.Round(notional, 2) + " USD)");
        if (notional >= MaxNotionalUsd)
        {
            Console.WriteLine("ABORT: notional >= " + MaxNotionalUsd + " USD safety cap.");
            return;
        }

        PredictionOrder? order = null;
        try
        {
            order = await exchange.CreateOrder(symbol, "limit", "buy", OrderSizeShares, price);
            Console.WriteLine("placed:  id " + order.Value.id + " | status " + order.Value.status);
            var fetched = await exchange.FetchOrder(order.Value.id, symbol);
            Console.WriteLine("fetched: id " + fetched.id + " | status " + fetched.status + " | remaining " + fetched.remaining);
        }
        finally
        {
            if (order != null && order.Value.id != null)
            {
                var canceled = await exchange.CancelOrder(order.Value.id, symbol);
                Console.WriteLine("canceled: id " + canceled.id + " | status " + canceled.status);
            }
        }
    }
}
